polkadot/xcm/xcm-builder/src/asset_exchange.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Adapters that implement the [`xcm_executor::traits::AssetExchange`] trait.
+
+use core::marker::PhantomData;
+use frame_support::traits::tokens::fungibles;
+use pallet_asset_conversion::{QuoteExchangePrice, SwapCredit};
+use sp_runtime::traits::Zero;
+use sp_std::vec;
+use xcm::prelude::*;
+use xcm_executor::{
+	traits::{AssetExchange, MatchesFungibles},
+	AssetsInHolding,
+};
+
+/// An adapter from [`pallet_asset_conversion::SwapCredit`] and
+/// [`pallet_asset_conversion::QuoteExchangePrice`] to [`xcm_executor::traits::AssetExchange`].
+///
+/// Takes just one fungible asset in `give` and allows only one fungible asset in `want`.
```

Comment:
The trait might allow multiple assets, but this particular type that implements that trait works in this way. Devs are free to use another type or build their own. I should make it clear that this is not a silver bullet though

Reply:
Then we should change the specification of `AssetExchange`.

It is literally written: `want: The minimum amount of assets which should be given to the caller in case any exchange happens.`

Or we can add a new method in `AssetExchange` like `exchange_one_asset`

Reply:
Also if `asset.len()` is 0 we can return OK

Reply:
You're right, I should error to make sure it doesn't fail silently

Reply:
I have to do tests for this type, so I'll cover this case as well

Reply:
Same with give